sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruning.scala

```diff
@@ -0,0 +1,153 @@
+          l @ LogicalRelation(hadoopFsRelation @ HadoopFsRelation(_, _,
+            dataSchema, _, parquetFormat: ParquetFileFormat, _), _, _, _)) =>
+        val projectionRootFields = projects.flatMap(getRootFields)
+        val filterRootFields = filters.flatMap(getRootFields)
+        val requestedRootFields = (projectionRootFields ++ filterRootFields).distinct
+
+        // If [[requestedRootFields]] includes a nested field, continue. Otherwise,
+        // return [[op]]
+        if (requestedRootFields.exists { case RootField(_, derivedFromAtt) => !derivedFromAtt }) {
+          val prunedSchema = requestedRootFields
+            .map { case RootField(field, _) => StructType(Array(field)) }
+            .reduceLeft(_ merge _)
+          val dataSchemaFieldNames = dataSchema.fieldNames.toSet
+          val prunedDataSchema =
+            StructType(prunedSchema.filter(f => dataSchemaFieldNames.contains(f.name)))
+
+          // If the data schema is different from the pruned data schema, continue. Otherwise,
+          // return [[op]]. We effect this comparison by counting the number of "leaf" fields in
+          // each schemata, assuming the fields in [[prunedDataSchema]] are a subset of the fields
+          // in [[dataSchema]].
+          if (countLeaves(dataSchema) > countLeaves(prunedDataSchema)) {
+            val prunedParquetRelation =
+              hadoopFsRelation.copy(dataSchema = prunedDataSchema)(hadoopFsRelation.sparkSession)
+
+            // We need to replace the expression ids of the pruned relation output attributes
+            // with the expression ids of the original relation output attributes so that
+            // references to the original relation's output are not broken
+            val outputIdMap = l.output.map(att => (att.name, att.exprId)).toMap
+            val prunedRelationOutput =
+              prunedParquetRelation
+                .schema
+                .toAttributes
+                .map {
+                  case att if outputIdMap.contains(att.name) =>
+                    att.withExprId(outputIdMap(att.name))
+                  case att => att
+                }
+            val prunedRelation =
+              l.copy(relation = prunedParquetRelation, output = prunedRelationOutput)
+
+            val projectionOverSchema = ProjectionOverSchema(prunedDataSchema)
+
+            // Construct a new target for our projection by rewriting and
+            // including the original filters where available
+            val projectionChild =
+              if (filters.nonEmpty) {
+                val projectedFilters = filters.map(_.transformDown {
+                  case projectionOverSchema(expr) => expr
+                })
+                val newFilterCondition = projectedFilters.reduce(And)
+                Filter(newFilterCondition, prunedRelation)
+              } else {
+                prunedRelation
+              }
+
+            // Construct the new projections of our [[Project]] by
```

Comment:
tiny nit: for inlined comment we could just better write `Project` or `` `Project` ``.

Reply:
No comment.